### PR TITLE
8287692: Move Class primitive APIs to jdk.internal.value.PrimitiveClass

### PR DESCRIPTION
**Pull request by RogerRiggs**
Move Class primitive APIs to jdk.internal.value.PrimitiveClass.

Update hotspot tests to use PrimitiveClass instead of Class.xxx methods.
Add imports of jdk.internal.value.PrimitiveClass
Add --add-exports java.base/jdk.internal.value=ALL-UNNAMED

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8287692](https://bugs.openjdk.org/browse/JDK-8287692): Move Class primitive APIs to jdk.internal.value.PrimitiveClass

### Reviewers
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - Committer) ⚠️ Review applies to [f9f1f7d4](https://git.openjdk.org/valhalla/pull/752/files/f9f1f7d4c7bfa97a53c3805a9981b8f6456efec5)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/valhalla pull/752/head:pull/752` \
`$ git checkout pull/752`

Update a local copy of the PR: \
`$ git checkout pull/752` \
`$ git pull https://git.openjdk.org/valhalla pull/752/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 752`

View PR using the GUI difftool: \
`$ git pr show -t 752`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/valhalla/pull/752.diff">https://git.openjdk.org/valhalla/pull/752.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/lang/Class.java [hunk 1]
```diff
@@ -203,6 +203,7 @@
     private static final int ANNOTATION = 0x00002000;
     private static final int ENUM       = 0x00004000;
     private static final int SYNTHETIC  = 0x00001000;
+    private static final int VALUE_CLASS     = 0x00000040;
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/Class.java:206]: Is this a merge error?  That was removed by JDK-8281463.

File: src/java.base/share/classes/java/lang/invoke/AbstractValidatingLambdaMetafactory.java [hunk 2]
```diff
@@ -268,7 +269,7 @@ void validateMetafactoryArgs() throws LambdaConversionException {
             }
 
             // check receiver type
-            if (!implClass.asPrimaryType().isAssignableFrom(receiverClass.asPrimaryType())) {
+            if (!PrimitiveClass.asPrimaryType(implClass).isAssignableFrom(PrimitiveClass.asPrimaryType(receiverClass))) {
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/invoke/AbstractValidatingLambdaMetafactory.java:272]: Maybe a convenient method `PrimitiveClass::isAssignableFrom`  is useful to add.

File: src/java.base/share/classes/java/lang/reflect/ProxyGenerator.java [hunk 3]
```diff
@@ -965,10 +967,11 @@ private void codeClassForName(MethodVisitor mv, Class<?> cl) {
                     "forName",
                     "(Ljava/lang/String;Z" + LJL_CLASSLOADER + ")Ljava/lang/Class;",
                     false);
-            if (cl.isPrimitiveValueType()) {
-              mv.visitMethodInsn(INVOKEVIRTUAL,
-                                 JL_CLASS,
-                                 "asValueType", "()Ljava/lang/Class;", false);
+            if (PrimitiveClass.isPrimitiveValueType(cl)) {
+              mv.visitMethodInsn(INVOKESTATIC,
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/reflect/ProxyGenerator.java:971]: nit: it's existing formatting issue.  use 4-space indentation.

File: src/java.base/share/classes/jdk/internal/access/JavaLangAccess.java [hunk 4]
```diff
@@ -525,16 +528,49 @@ public interface JavaLangAccess {
...
      */
     void unparkVirtualThread(Thread thread);
 
     /**
      * Creates a new StackWalker
      */
-    StackWalker newStackWalkerInstance(Set<StackWalker.Option> options,
-                                       ContinuationScope contScope,
-                                       Continuation continuation);
+    StackWalker newStackWalkerInstance(Set<StackWalker.Option> options, ContinuationScope contScope, Continuation continuation);
```
Comment (mlchung) [inline src/java.base/share/classes/jdk/internal/access/JavaLangAccess.java:541]: Is the formatting change in this file (this line and above) accidental done by IDE?

I expect the change should just be the addition of the new methods below.

File: src/java.base/share/classes/jdk/internal/value/PrimitiveClass.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+import java.lang.reflect.Method;
+
+/**
+ * Utilities to access Primitive Classes as described in JEP 401.
+ */
+public class PrimitiveClass {
+
+    // ACC_PRIMITIVE from JEP 401
+    private static final int PRIMITIVE_CLASS = 0x00000800;
```
Comment (mlchung) [inline src/java.base/share/classes/jdk/internal/value/PrimitiveClass.java:39]: perhaps make this public to have a single place in the libraries to define this constant and remove `Class::PRIMITIVE_CLASS`.

**Review by mlchung — COMMENTED** [review]
Looks okay in general.

File: test/jdk/valhalla/valuetypes/ProxyTest.java [hunk 6]
```diff
@@ -30,6 +30,7 @@
 import java.lang.reflect.*;
 import java.util.Arrays;
 
+
```
Comment (mlchung) [inline test/jdk/valhalla/valuetypes/ProxyTest.java:33]: No change in this file.

File: test/jdk/java/lang/invoke/condy/BootstrapMethodJumboArgsTest.java [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2017, 2022, Oracle and/or its affiliates. All rights reserved.
```
Comment (mlchung) [inline test/jdk/java/lang/invoke/condy/BootstrapMethodJumboArgsTest.java:2]: is this change needed?

File: test/jdk/valhalla/valuetypes/UninitializedValueTest.java [hunk 8]
```diff
@@ -35,6 +35,9 @@
 import java.lang.invoke.MethodHandles;
 import java.lang.reflect.Field;
 
+import jdk.internal.value.PrimitiveClass;
+
+
```
Comment (mlchung) [inline test/jdk/valhalla/valuetypes/UninitializedValueTest.java:40]: extra new line

File: test/jdk/valhalla/valuetypes/ObjectMethodsViaCondy.java [hunk 9]
```diff
@@ -24,8 +24,9 @@
 /*
  * @test
  * @summary Test ObjectMethods::bootstrap call via condy
+ * @modules java.base/jdk.internal.value:+open
```
Comment (mlchung) [inline test/jdk/valhalla/valuetypes/ObjectMethodsViaCondy.java:27]: Is `+open` needed?   PrimitiveClass methods are all public.  `+open` is needed only for deep reflection.

Same comments apply to other tests.

**Comment by RogerRiggs** [comment]
@mchung Thanks for the review and suggestions.

File: src/jdk.unsupported/share/classes/sun/misc/Unsafe.java [hunk 10]
```diff
@@ -651,7 +651,7 @@ public long objectFieldOffset(Field f) {
         if (declaringClass.isHidden()) {
             throw new UnsupportedOperationException("can't get field offset on a hidden class: " + f);
         }
-        if (f.getDeclaringClass().isPrimitiveClass()) {
+        if (PrimitiveClass.isPrimitiveClass(f.getDeclaringClass())) {
             throw new UnsupportedOperationException("can't get field offset on an inline class: " + f);
```
Comment (mlchung) [inline src/jdk.unsupported/share/classes/sun/misc/Unsafe.java:655]: since you are in this file, can you update the error message `s/inline class/primitive class/`

File: test/hotspot/jtreg/serviceability/jvmti/Valhalla/HeapDump/HeapDump.java [hunk 11]
```diff
@@ -264,9 +264,10 @@ private static void compareObjectFields(String logPrefix, Object testObj, JavaOb
         Field[] fields = testObj.getClass().getDeclaredFields();
         for (Field testField : fields) {
             boolean isStatic = Modifier.isStatic(testField.getModifiers());
+            boolean isPrimitive = (testField.getModifiers() & 0x0800) != 0; // TBD: until the modifiers settle down
```
Comment (mlchung) [inline test/hotspot/jtreg/serviceability/jvmti/Valhalla/HeapDump/HeapDump.java:267]: This can be replaced with `PrimitiveClass.isPrimitiveClass`

File: test/jdk/valhalla/valuetypes/ProxyTest.java (hunk 6 above)
Comment (mlchung) [inline test/jdk/valhalla/valuetypes/ProxyTest.java:33]: This test should have no modification.   This extra line can be cleaned up.

**Review by mlchung — APPROVED** [review]
JavaLangAccess.java has whitespace changes which I believe it's not intentional.  Please revert those whitespace change before pushing the changes.

File: src/java.base/share/classes/jdk/internal/value/PrimitiveClass.java [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+public class PrimitiveClass {
+
+    /**
+     * ACC_PRIMITIVE modifier defined by JEP 401. Subject to change.
+     */
+    public static final int PRIMITIVE_CLASS = 0x00000800;
+
+    private static final JavaLangAccess javaLangAccess = SharedSecrets.getJavaLangAccess();
+
+
```
Comment (ExE-Boss) [inline src/java.base/share/classes/jdk/internal/value/PrimitiveClass.java:43]: excess newline

File: src/java.base/share/classes/jdk/internal/value/PrimitiveClass.java [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+     * @param aClass a class
+     * @return {@code true} if this class is a primitive class, otherwise {@code false}
+     * @see Class#isValue()
+     * @see #asPrimaryType(Class)
+     * @see #asValueType(Class)
+     * @since Valhalla
+     */
+    public static boolean isPrimitiveClass(Class<?> aClass) {
+        return javaLangAccess.isPrimitiveClass(aClass);
+
```
Comment (ExE-Boss) [inline src/java.base/share/classes/jdk/internal/value/PrimitiveClass.java:145]: excess newline
